From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Matt Carroll <matthew-carroll@users.noreply.github.com>
Date: Sun, 3 Nov 2024 14:49:01 -0800
Subject: Added gclient files to gitignore to support Flock and other Flutter
 forks.

diff --git a/.gitignore b/.gitignore
index 73728cebbfccffd6864ddf5f7d46a3c8f4fffa38..7eb4ad199e0b8515f896ce9dbc91d437292bc2a3 100644
--- a/.gitignore
+++ b/.gitignore
@@ -1,3 +1,8 @@
+# Injected by Flock to support custom forks
+.gclient
+.gclient_entries
+.gclient_previous_sync_commits
+
 # Do not remove or rename entries in this file, only add new ones
 # See https://github.com/flutter/flutter/issues/128635 for more context.
 
