From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jesse Ezell <jezell@users.noreply.github.com>
Date: Sun, 2 June 2025 14:49:01 -0800
Subject: Fix iOS builds

diff --git a/engine/src/flutter/shell/platform/darwin/common/framework/Source/Logger.swift b/engine/src/flutter/shell/platform/darwin/common/framework/Source/Logger.swift
index 089b2068a3..2d0f7db8ba 100644
--- a/engine/src/flutter/shell/platform/darwin/common/framework/Source/Logger.swift
+++ b/engine/src/flutter/shell/platform/darwin/common/framework/Source/Logger.swift
@@ -120,7 +120,10 @@ final class SyslogOutputWriter: OutputWriter {
   func writeLine(level: LogLevel, _ message: String) {
     // TODO(cbracken): replace this with os_log-based approach.
     // https://github.com/flutter/flutter/issues/44030
-    message.withCString { vsyslog(LOG_ALERT, "%s", getVaList([$0])) }
+    // Build a C `va_list` from the Swift string and pass it to vsyslog.
+    withVaList([message]) { vaList in
+      vsyslog(LOG_ALERT, "%s", vaList)
+    }
   }
 }
